### PR TITLE
Improve jsonschema validation error messages

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

Right now the jsonschema validation errors are next to useless because
of how the schemas files our constructed. The schemas are built in away
that tries to constrict too much of the structure that uses oneOf in
multiple levels of the schema. These oneOf's make the schema much more
difficult to debug because any rule that fails in the subschema below
those will not actually be what fails validation, but instead it will be
the oneOf lines because none of the elements in the list of valid
subschemas passed validation. This becomes very problematic for
debugging things that get jsonschema validated, especially with our
current error message structure. The current error message prints the
entire error message from jsonschema to either the exception message if
it's less than 1000 characters (which it rarely is) or as a debug log
message to python logging if it's more than 1000 characters. These
messages are not very useful since the contain most of the schema as
the rule that failed (which is why it's often over 1000 characters)
because of the use of oneOf checks.

This commit changes our error reporting to try and be a bit more useful.
It leverages the best_match [1] feature from jsonschema to use a
heuristic to try and figure out what actually caused the validation
failure instead of just blindly saying it was the whole schema. In
local testing this often gets the right spot on the input dictionary
while not necessarily the exact rule. For example, if you passed an
instruction for a qasm experiment missing a qubits field it says the
name field was invalid for that instruction. While not exact this is a
much closer hint for debugging. The error messages are adjusted to use
the best match path to point to the element (not the specific field)
that failed validation, and then print the error message from best match
and saying it's possibly the error. This should make it easier to find
errors in qobj dictionary that isn't passing schema validation. It's not
always perfect, but something more targetted should hopefully be better
than everything, which is the status quo.

A follow-up here will be to try and rewrite the schema to not rely on so
many high level oneOf rules. This will likely result in relaxing the
schema slightly so validation won't catch high level errors (like
sending both a qasm configuration and a pulse configuration at the same
time). But, those errors should be next to non-existent and are not
really the class of errors we should be trying to validate with schema
anyway. Especially for qiskit where the qobj dictionaries are being
generated and we don't have to worry about such cases.

### Details and comments

[1] https://python-jsonschema.readthedocs.io/en/stable/errors/#best-match-and-relevance

### CODE REVIEW

**Comment by taalexander** [comment]
It looks like a good improvement! Do you have an example of the error output now?

**Comment by mtreinish** [comment]
Sure, here are some of my examples from testing. An empty dict

`validate_json_against_schema({}, 'qobj')`

goes from:

```
qiskit.validation.jsonschema.exceptions._SummaryValidationError: "Original message too long to be useful: {} is not valid under any of the given schemas\n\nFailed validating 'oneOf' in schema:\n    {'$schema': 'http://json-schema.org/draft-04/schema#',\n     'definitions': {'bfunc': {'properties': {'mask': {'type': 'string'},\n                                              'memory': {'minimum': 0,\n                                                         'type': 'integer'},\n                                              'name': {'enum': ['bfunc']},\n                                              'register': {'minimum': 0,\n                                                           'type': 'integer'},\n                                              'relation': {'type': 'string'},\n                                              'val': {'type': 'string'}},\n                               'required': ['mask', 'relation', 'val'],\n                               'type': 'object'},\n                     'complex': {'description': 'JSON representation of a '\n                                                'complex [...]"

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "test.py", line 30, in <module>
    validation.jsonschema.validate_json_against_schema({}, 'qobj')
  File "/home/computertreker/git/qiskit/qiskit-core/qiskit/validation/jsonschema/schema_validation.py", line 149, in validate_json_against_schema
    raise newerr
qiskit.validation.jsonschema.exceptions.SchemaValidationError: 'JSON failed validation. Set Qiskit log level to DEBUG for further information.'
```
without this PR to:

```
Traceback (most recent call last):
  File "test.py", line 30, in <module>
    validation.jsonschema.validate_json_against_schema({}, 'qobj')
  File "/home/computertreker/git/qiskit/qiskit-core/qiskit/validation/jsonschema/schema_validation.py", line 156, in validate_json_against_schema
    raise SchemaValidationError(err_message)
qiskit.validation.jsonschema.exceptions.SchemaValidationError: "Validation failed. Possibly because 'qobj_id' is a required property"
```

**Comment by mtreinish** [comment]
Taking an qobj from assemble and messing with a field like:

```
qc = random_circuit(2, 2, measure=True, conditional=True, seed=512)
qobj = assemble(circuits, qobj_id='1024', backend=backend).to_dict()
qobj['experiments'][0]['instructions'][2]['name'] = 'measure'
validate_json_against_schema(qobj, 'qobj')
```

Goes from:

```
qiskit.validation.jsonschema.exceptions._SummaryValidationError: "Original message too long to be useful: {'qobj_id': '1024', 'config': {'n_qubits': 2, 'memory_slots': 2, 'shots': 1024, 'fast': True, 'parameter_binds': [], 'memory': False}, 'schema_version': '1.1.0', 'type': 'QASM', 'experiments': [{'config': {'n_qubits': 2, 'memory_slots': 2}, 'instructions': [{'name': 'cz', 'qubits': [0, 1]}, {'name': 'u3', 'qubits': [1], 'params': [1.4157922722110299, 3.5132251685231872, 4.287246470451548]}, {'name': 'measure', 'qubits': [0]}, {'memory': [0], 'name': 'measure', 'qubits': [0]}, {'memory': [1], 'name': 'measure', 'qubits': [1]}], 'header': {'creg_sizes': [['c', 2]], 'memory_slots': 2, 'n_qubits': 2, 'qubit_labels': [['q', 0], ['q', 1]], 'clbit_labels': [['c', 0], ['c', 1]], 'qreg_sizes': [['q', 2]], 'name': 'circuit0'}}], 'header': {'backend_version': '0.0.0', 'backend_name': 'fake_almaden'}} is not valid under any of the given schemas\n\nFailed validating 'oneOf' in schema:\n    {'$schema': 'http://json-schema.org/draft-04/schema#',\n     'definitions': {'bfunc': {'properties': {'mask': {'ty[...]"

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "test.py", line 30, in <module>
    validation.jsonschema.validate_json_against_schema(fqobj_dict, 'qobj')
  File "/home/computertreker/git/qiskit/qiskit-core/qiskit/validation/jsonschema/schema_validation.py", line 149, in validate_json_against_schema
    raise newerr
qiskit.validation.jsonschema.exceptions.SchemaValidationError: 'JSON failed validation. Set Qiskit log level to DEBUG for further information.'
```

to

```
Traceback (most recent call last):
  File "test.py", line 30, in <module>
    validation.jsonschema.validate_json_against_schema(fqobj_dict, 'qobj')
  File "/home/computertreker/git/qiskit/qiskit-core/qiskit/validation/jsonschema/schema_validation.py", line 156, in validate_json_against_schema
    raise SchemaValidationError(err_message)
qiskit.validation.jsonschema.exceptions.SchemaValidationError: "Validation failed. Possibly at ['experiments'][0]['instructions'][2] because of {'enum': ['u3', 'U', 'id', 'h', 's', 'sdg', 't', 'tdg', 'x', 'y', 'z', 'u1', 'u0', 'rx', 'ry', 'rz', 'u2', 'cx', 'CX', 'cy', 'cz', 'ch', 'swap', 'cu1', 'crz', 'rzz', 'cu3', 'ccx', 'cswap', 'snapshot', 'reset', 'barrier', 'bfunc', 'copy', 'measure']} is not allowed for 'measure'"
```

with this pr. Which is the right instruction but not the correct rule on the instruction.

**Comment by mtreinish** [comment]
There are some cases where the guess is completely off the mark, normally in something like run_config. For example:

```
qc = random_circuit(2, 2, measure=True, conditional=True, seed=512)
qobj = assemble(circuits, qobj_id='1024', backend=backend).to_dict()
qobj['config'] = {'shots': 2204.88}
validate_json_against_schema(qobj, 'qobj')
```

Goes from without this PR:

```
qiskit.validation.jsonschema.exceptions._SummaryValidationError: "Original message too long to be useful: {'schema_version': '1.1.0', 'experiments': [{'header': {'memory_slots': 2, 'qubit_labels': [['q', 0], ['q', 1]], 'name': 'circuit0', 'qreg_sizes': [['q', 2]], 'clbit_labels': [['c', 0], ['c', 1]], 'n_qubits': 2, 'creg_sizes': [['c', 2]]}, 'instructions': [{'name': 'cz', 'qubits': [0, 1]}, {'name': 'u3', 'params': [1.4157922722110299, 3.5132251685231872, 4.287246470451548], 'qubits': [1]}, {'name': 't', 'qubits': [0]}, {'name': 'measure', 'memory': [0], 'qubits': [0]}, {'name': 'measure', 'memory': [1], 'qubits': [1]}], 'config': {'memory_slots': 2, 'n_qubits': 2}}], 'type': 'QASM', 'qobj_id': '1024', 'config': {'shots': 2204.88}, 'header': {'backend_name': 'fake_almaden', 'backend_version': '0.0.0'}} is not valid under any of the given schemas\n\nFailed validating 'oneOf' in schema:\n    {'$schema': 'http://json-schema.org/draft-04/schema#',\n     'definitions': {'bfunc': {'properties': {'mask': {'type': 'string'},\n                                              'memory': {'minimum': 0,\n    [...]"

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "test.py", line 31, in <module>
    validation.jsonschema.validate_json_against_schema(fqobj_dict, 'qobj')
  File "/home/computertreker/git/qiskit/qiskit-core/qiskit/validation/jsonschema/schema_validation.py", line 149, in validate_json_against_schema
    raise newerr
qiskit.validation.jsonschema.exceptions.SchemaValidationError: 'JSON failed validation. Set Qiskit log level to DEBUG for further information.'
```

to

```
Traceback (most recent call last):
  File "test.py", line 31, in <module>
    validation.jsonschema.validate_json_against_schema(fqobj_dict, 'qobj')
  File "/home/computertreker/git/qiskit/qiskit-core/qiskit/validation/jsonschema/schema_validation.py", line 156, in validate_json_against_schema
    raise SchemaValidationError(err_message)
qiskit.validation.jsonschema.exceptions.SchemaValidationError: "Validation failed. Possibly at ['experiments'][0]['instructions'][0] because of 'cz' is not one of ['pv']"
```
with this pr. Which is not even close to the failure.

**Comment by taalexander** [comment]
It definitely looks better for the first two cases. I suppose the third case could be very frustrating if encountered.

**Comment by mtreinish** [comment]
I don't want to hold up the release over this, it's a minor debugging thing and is only really needed on master so I'm moving the milestone back to 0.13

File: qiskit/validation/jsonschema/schema_validation.py [hunk 1]
```diff
@@ -131,21 +131,38 @@ def validate_json_against_schema(json_dict, schema,
...
-        newerr = SchemaValidationError(err_msg)
-        newerr.__cause__ = _SummaryValidationError(err)
-        logger.debug('%s', _format_causes(err))
+        except jsonschema.ValidationError as err:
+            if err_msg is None:
+                err_msg = ("JSON failed validation. Set Qiskit log level to "
+                           "DEBUG for further information.")
+            newerr = SchemaValidationError(err_msg)
+            newerr.__cause__ = _SummaryValidationError(err)
         raise newerr
```
Comment (taalexander) [inline qiskit/validation/jsonschema/schema_validation.py]: Not blocking but would this be more python if it was
```python
raise newerr from _SummaryValidationError(err)
```
?

Comment (mtreinish) [reply qiskit/validation/jsonschema/schema_validation.py]: Sure, I was just indenting what was there before to minimize the changes. I can update it if needed though.

**Pull request merged**